DKGRituals/src/mappingCoordinator.ts

```diff
@@ -0,0 +1,137 @@
+  StartAggregationRound as StartAggregationRoundEvent,
+  AggregationPosted as AggregationPostedEvent,
+  EndRitual as EndRitualEvent,
+  RitualAuthorityTransferred as RitualAuthorityTransferredEvent,
+} from "../generated/Coordinator/Coordinator";
+import {
+  getOrCreateRitual,
+  getOrCreateRitualCounter,
+  getOrCreateTransaction,
+  getOrCreateUser,
+} from "./utils/helper";
+import { RITUAL_COUNTER_ID } from "./utils/constants";
+import { calculatePublicKey, getTransactionIDFromEvent } from "./utils/utils";
+
+export function handleStartRitual(event: StartRitualEvent): void {
+  const coordinatorContract = Coordinator.bind(event.address);
+  const ritualInst = coordinatorContract.rituals(event.params.ritualId);
+
+  const transaction = getOrCreateTransaction(getTransactionIDFromEvent(event));
+  transaction.txHash = event.transaction.hash;
+  transaction.timestamp = event.block.timestamp;
+  transaction.from = event.transaction.from;
+  transaction.to = event.transaction.to;
+  transaction.description = "Start Ritual";
+  transaction.save();
+
+  const ritual = getOrCreateRitual(event.params.ritualId.toString());
+  ritual.initiator = ritualInst.getInitiator();
+  ritual.initTimestamp = ritualInst.getInitTimestamp();
+  ritual.endTimestamp = ritualInst.getEndTimestamp();
+  ritual.authority = event.params.authority;
+  ritual.dkgSize = ritualInst.getDkgSize();
+  ritual.threshold = ritualInst.getThreshold();
+  ritual.accessController = ritualInst.getAccessController();
+  ritual.participants = changetype<Bytes[]>(event.params.participants);
+  ritual.dkgStatus = "DKG_AWAITING_TRANSCRIPTS";
+
+  const transactions = ritual.transactions;
+  transactions.push(transaction.id);
+  ritual.transactions = transactions;
+  ritual.save();
+
+  const ritualCounter = getOrCreateRitualCounter(RITUAL_COUNTER_ID);
+  ritualCounter.total = ritualCounter.total + 1;
+  ritualCounter.notEnded = ritualCounter.notEnded + 1;
+  ritualCounter.save();
+
+  const user = getOrCreateUser(ritual.initiator);
+  const rituals = user.rituals;
+  rituals.push(ritual.id);
+  user.rituals = rituals;
+  user.save();
+}
+
+export function handleTranscriptPosted(event: TranscriptPostedEvent): void {
+  const ritual = getOrCreateRitual(event.params.ritualId.toString());
+  const postedTranscripts = ritual.postedTranscripts;
+  postedTranscripts.push(event.params.node);
+  ritual.postedTranscripts = postedTranscripts;
+  ritual.postedTranscriptsAmount = ritual.postedTranscriptsAmount + 1;
+  ritual.save();
+}
+
+export function handleStartAggregationRound(
+  event: StartAggregationRoundEvent,
+): void {
+  const transaction = getOrCreateTransaction(getTransactionIDFromEvent(event));
+  transaction.txHash = event.transaction.hash;
+  transaction.timestamp = event.block.timestamp;
+  transaction.from = event.transaction.from;
+  transaction.to = event.transaction.to;
+  transaction.description = "Posted Transcripts";
+  transaction.save();
```

Comment:
Do you need this? This event is emitted from the Coordinator contract when the last node to post their transcript does so -https://github.com/nucypher/nucypher-contracts/blob/main/contracts/contracts/coordination/Coordinator.sol#L340. So the tx associated with this event is presumably the same as the last `TranscriptPosted` event. Would this cause the tx to be duplicated in the list of transactions?

Reply:
Transactions entity is needed here precisely to capture the transaction details from the last tx that would trigger this event.
Essentially It is capturing the tx so we can have a chronological order of transactions that trigger the change of state of the ritual along time. Each ritual has a transactions property and only the tx that triggers the change of state are added into the transactions list of the Ritual entity as you can see it here: https://api.studio.thegraph.com/proxy/59264/taco-matic/version/latest/graphql?query=query+MyQuery+%7B%0A++rituals%28first%3A+10%2C+orderBy%3A+initTimestamp%29+%7B%0A++++id%0A++++publicKey%0A++++transactions%28first%3A+10%2C+orderBy%3A+timestamp%29+%7B%0A++++++timestamp%0A++++++description%0A++++++id%0A++++%7D%0A++++dkgStatus%0A++%7D%0A%7D